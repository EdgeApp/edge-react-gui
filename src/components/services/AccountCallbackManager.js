// @flow

import { type EdgeAccount, type EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { updateExchangeRates } from '../../actions/ExchangeRateActions.js'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions.js'
import { newTransactionsRequest, refreshTransactionsRequest } from '../../actions/TransactionListActions.js'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions.js'
import { useAsyncEffect } from '../../hooks/useAsyncEffect.js'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber.js'
import { useEffect, useState } from '../../types/reactHooks.js'
import { useDispatch } from '../../types/reactRedux.js'
import { type NavigationProp, useNavigation } from '../../types/routerTypes.js'
import { isReceivedTransaction, snooze } from '../../util/utils.js'
import { WcSmartContractModal } from '../modals/WcSmartContractModal.js'
import { Airship } from './AirshipInstance.js'

type Props = {
  account: EdgeAccount
}

// Tracks items that need a refresh:
type DirtyList = {
  rates: boolean,
  walletList: boolean,
  wallets: { [walletId: string]: EdgeCurrencyWallet }
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false,
  wallets: {}
}

export function AccountCallbackManager(props: Props) {
  const { account } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = useState<DirtyList>(notDirty)
  const navigation: NavigationProp<'edge'> = useNavigation()

  // Helper for marking wallets dirty:
  function addWallet(wallet: EdgeCurrencyWallet) {
    setDirty(dirty => ({
      ...dirty,
      rates: true, // We might reconsider this behavior
      wallets: { ...dirty.wallets, [wallet.id]: wallet }
    }))
  }

  // Subscribe to the account:
  useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (account.loggedIn === false) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        navigation.push('securityAlerts')
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        dispatch(newTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (isReceivedTransaction(transactions[finalTxIndex])) {
          dispatch(checkPasswordRecovery())
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)

        dispatch(refreshTransactionsRequest(wallet.id, transactions))
        addWallet(wallet)
      }),

      wallet.on('wcNewContractCall', obj => {
        const { dApp, payload, uri, walletId } = obj
        if (walletId == null) return
        Airship.show(bridge => <WcSmartContractModal bridge={bridge} walletId={walletId} dApp={dApp} payload={payload} uri={uri} />)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => addWallet(wallet)),
      wallet.watch('enabledTokenIds', () => addWallet(wallet)),
      wallet.watch('name', () => addWallet(wallet))
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(async () => {
    setDirty(notDirty)

    // Update wallets:
    const walletIds = Object.keys(dirty.wallets)
    if (dirty.walletList) {
      // Update all wallets (hammer mode):
      console.log('Updating wallet list')
      await dispatch(updateWalletsRequest())
      await snooze(1000)
    } else if (walletIds.length > 0) {
      // Update individual wallets:
      console.log(`Updating wallets: ${walletIds.map(id => walletPrefix(dirty.wallets[id])).join(' ')}`)
      dispatch({
        type: 'UI/WALLETS/UPSERT_WALLETS',
        data: {
          wallets: walletIds.map(id => dirty.wallets[id])
        }
      })
      await snooze(1000)
    }

    // Update exchange rates:
    if (dirty.rates) {
      console.log('Updating exchange rates')
      await dispatch(updateExchangeRates())
      await snooze(1000)
    }
  }, [dirty])

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
