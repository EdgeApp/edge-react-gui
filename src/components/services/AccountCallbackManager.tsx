import { EdgeAccount, EdgeCurrencyWallet } from 'edge-core-js'
import { watchSecurityAlerts } from 'edge-login-ui-rn'
import * as React from 'react'

import { showBackupModal } from '../../actions/BackupModalActions'
import { updateExchangeRates } from '../../actions/ExchangeRateActions'
import { checkFioObtData } from '../../actions/FioActions'
import { showReceiveDropdown } from '../../actions/ReceiveDropdown'
import { checkPasswordRecovery } from '../../actions/RecoveryReminderActions'
import { updateWalletLoadingProgress, updateWalletsRequest } from '../../actions/WalletActions'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useWalletsSubscriber } from '../../hooks/useWalletsSubscriber'
import { stakeMetadataCache } from '../../plugins/stake-plugins/metadataCache'
import { useDispatch } from '../../types/reactRedux'
import { NavigationBase } from '../../types/routerTypes'
import { snooze } from '../../util/utils'
import { Airship } from './AirshipInstance'

interface Props {
  account: EdgeAccount
  navigation: NavigationBase
}

// Tracks items that need a refresh:
interface DirtyList {
  rates: boolean
  walletList: boolean
}

const notDirty: DirtyList = {
  rates: false,
  walletList: false
}

export function AccountCallbackManager(props: Props) {
  const { account, navigation } = props
  const dispatch = useDispatch()
  const [dirty, setDirty] = React.useState<DirtyList>(notDirty)

  // Helper for marking wallets dirty:
  function setRatesDirty() {
    setDirty(dirty => ({
      ...dirty,
      rates: true
    }))
  }

  // Subscribe to the account:
  React.useEffect(() => {
    const cleanups = [
      account.watch('currencyWallets', () =>
        setDirty(dirty => ({
          ...dirty,
          walletList: true
        }))
      ),

      account.watch('loggedIn', () => {
        if (!account.loggedIn) {
          Airship.clear()
          console.log('onLoggedOut')
        }
      }),

      watchSecurityAlerts(account, hasAlerts => {
        if (hasAlerts) {
          navigation.navigate('securityAlerts', {})
        }
      }),

      account.rateCache.on('update', () =>
        setDirty(dirty => ({
          ...dirty,
          rates: true
        }))
      )
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  }, [account, navigation])

  // Subscribe to each wallet that comes online:
  useWalletsSubscriber(account, wallet => {
    const cleanups = [
      wallet.watch('syncRatio', ratio => {
        dispatch(updateWalletLoadingProgress(wallet.id, ratio))
      }),

      wallet.on('newTransactions', transactions => {
        for (const tx of transactions) {
          const txid = tx.txid.toLowerCase()
          const cacheEntries = stakeMetadataCache[txid]
          // Assign cached stake metadata
          if (cacheEntries != null) {
            cacheEntries.forEach(cacheEntry => {
              const { currencyCode, metadata } = cacheEntry
              if (tx.currencyCode !== currencyCode) return
              wallet.saveTx({ ...tx, metadata }).catch(err => console.warn(err))
            })

            delete stakeMetadataCache[txid]
          }
        }

        console.log(`${walletPrefix(wallet)}: onNewTransactions: ${transactions.map(tx => tx.txid).join(' ')}`)

        // Check for incoming FIO requests:
        const receivedTxs = transactions.filter(tx => !tx.isSend)
        if (receivedTxs.length > 0) dispatch(checkFioObtData(wallet, receivedTxs)).catch(err => console.warn(err))

        // Show the dropdown for the first transaction:
        const [firstReceive] = receivedTxs
        if (firstReceive != null) {
          dispatch(showReceiveDropdown(navigation, firstReceive))

          // Notify the user to consider backing up their account
          if (account.username == null) {
            showBackupModal({ navigation })
          }
        }

        // Check if password recovery is set up:
        const finalTxIndex = transactions.length - 1
        if (!transactions[finalTxIndex].isSend) {
          dispatch(checkPasswordRecovery(navigation))
        }
      }),

      wallet.on('transactionsChanged', transactions => {
        console.log(`${walletPrefix(wallet)}: onTransactionsChanged: ${transactions.map(tx => tx.txid).join(' ')}`)
      }),

      // These ones defer their work until later:
      wallet.watch('balances', () => setRatesDirty()),
      wallet.watch('enabledTokenIds', () => setRatesDirty())
    ]

    return () => cleanups.forEach(cleanup => cleanup())
  })

  // Do the expensive work with rate limiting:
  useAsyncEffect(
    async () => {
      setDirty(notDirty)

      // Update wallets:
      if (dirty.walletList) {
        // Update all wallets (hammer mode):
        console.log('Updating wallet list')
        await dispatch(updateWalletsRequest())
        await snooze(1000)
      }

      // Update exchange rates:
      if (dirty.rates) {
        console.log('Updating exchange rates')
        await dispatch(updateExchangeRates())
        await snooze(1000)
      }
    },
    [dirty],
    'AccountCallbackManager'
  )

  return null
}

function walletPrefix(wallet: EdgeCurrencyWallet): string {
  return `${wallet.currencyInfo.currencyCode}-${wallet.id.slice(0, 2)}`
}
