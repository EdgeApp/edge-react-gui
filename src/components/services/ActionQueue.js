// @flow

import { type EdgeAccount } from 'edge-core-js'

import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueue = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  // TODO: Fix sync server stale ActionQueue state overwrite on startup
  //
  // Initialization
  //

  // useAsyncEffect(async () => {
  //   if (account?.dataStore != null) {
  //     const store = makeActionQueueStore(account)
  //     const queue = await store.getActionQueueMap()
  //     dispatch({
  //       type: 'ACTION_QUEUE/LOAD_QUEUE',
  //       data: queue
  //     })
  //   }
  // }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(
          programId =>
            // Ignore running programs
            !executing[programId]
        )
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]
          const { nextState } = await executeActionProgram(account, program, state).catch((error: Error) => ({
            nextState: {
              ...state,
              effect: { type: 'done', error }
            }
          }))

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
