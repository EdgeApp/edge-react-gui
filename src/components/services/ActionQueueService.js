// @flow

import { type EdgeAccount } from 'edge-core-js'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { mockActionProgram } from '../../controllers/action-queue/mock'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime'
import { type ActionQueueMap, type ExecutionContext } from '../../controllers/action-queue/types'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useRef } from '../../types/reactHooks'
import { useDispatch, useSelector } from '../../types/reactRedux'

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const queue: ActionQueueMap = useSelector(state => state.actionQueue.queue)
  const executingRef = useRef<{ [programId: string]: boolean }>({})

  const executionContext: ExecutionContext = { account, clientId }

  //
  // Initialization
  //

  useAsyncEffect(async () => {
    if (account?.dataStore != null) {
      const store = makeActionQueueStore(account, clientId)
      const queue = await store.getActionQueueMap()
      dispatch({
        type: 'ACTION_QUEUE/LOAD_QUEUE',
        data: queue
      })
    }
  }, [account, dispatch])

  //
  // Runtime
  //

  useAsyncEffect(async () => {
    if (queue != null) {
      const executing = executingRef.current
      const promises = Object.keys(queue)
        .filter(programId => !executing[programId] && queue[programId].state.clientId === clientId)
        .map(async programId => {
          // Set program to running
          executing[programId] = true

          const { program, state } = queue[programId]

          if (program.mockMode) {
            const { nextState } = await mockActionProgram(account, program, state)
            dispatch(updateActionProgramState(nextState))
            executing[programId] = false
            return
          }

          const { nextState } = await executeActionProgram(executionContext, program, state).catch((error: Error) => {
            console.warn(new Error('Action Program Exception: ' + error.message))
            console.error(error)
            return {
              nextState: {
                ...state,
                effect: { type: 'done', error }
              }
            }
          })

          // Update program state
          dispatch(updateActionProgramState(nextState))

          // Unset program to running
          executing[programId] = false
        })

      await Promise.all(promises)
    }
  }, [queue])

  return null
}
