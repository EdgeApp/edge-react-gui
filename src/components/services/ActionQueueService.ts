import { EdgeAccount } from 'edge-core-js'
import * as React from 'react'
import { useRef } from 'react'

import { makeActionQueueStore } from '../../controllers/action-queue/ActionQueueStore'
import { asCleanError } from '../../controllers/action-queue/cleaners'
import { updateActionProgramState } from '../../controllers/action-queue/redux/actions'
import { executeActionProgram } from '../../controllers/action-queue/runtime/executeActionProgram'
import { ActionProgramState, ActionQueueMap, ExecutionResults } from '../../controllers/action-queue/types'
import { wasPushRequestBody } from '../../controllers/action-queue/types/pushApiTypes'
import { useAsyncEffect } from '../../hooks/useAsyncEffect'
import { useExecutionContext } from '../../hooks/useExecutionContext'
import { useHandler } from '../../hooks/useHandler'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { makePeriodicTask } from '../../util/PeriodicTask'
import { makePushClient } from '../../util/PushClient/PushClient'

const EXECUTION_INTERVAL = 1000

interface ServiceProgramStates {
  [programId: string]: {
    executing: boolean
  }
}

export const ActionQueueService = () => {
  const dispatch = useDispatch()
  const account: EdgeAccount = useSelector(state => state.core.account)
  const clientId: string = useSelector(state => state.core.context.clientId)
  const actionQueueMap: ActionQueueMap = useSelector(state => state.actionQueue.actionQueueMap)
  const activeProgramIds = useSelector(state => state.actionQueue.activeProgramIds)
  const serviceProgramStatesRef = useRef<ServiceProgramStates>({})

  const executionContext = useExecutionContext()
  const executionContextMock = useExecutionContext(true)

  const updateProgramState = useHandler(async (state: ActionProgramState, executing: boolean) => {
    const { programId } = state
    // Update service level state
    const serviceProgramStates = serviceProgramStatesRef.current
    serviceProgramStates[programId] = { ...serviceProgramStates[programId], executing }
    // Update action queue level state
    await dispatch(
      updateActionProgramState({
        ...state,
        executing: executing
      })
    )
  })

  //
  // Log Dev Info
  //
  React.useEffect(() => {
    if (__DEV__ && account != null && account.rootLoginId != null && clientId != null) {
      const pushClient = makePushClient(account, clientId)
      const requestBody = pushClient.getPushRequestBody()
      console.log('***********************')
      console.log('PUSH SERVER DEV INFO:', wasPushRequestBody(requestBody))
      console.log('***********************')
    }
  }, [account, clientId])

  //
  // Initialization
  //

  useAsyncEffect(
    async () => {
      if (account?.dataStore != null) {
        const store = makeActionQueueStore(account, clientId)
        const queue = await store.getActionQueueMap()
        dispatch({
          type: 'ACTION_QUEUE/LOAD_QUEUE',
          data: queue
        })
      }
    },
    [account, dispatch],
    'ActionQueueService'
  )

  //
  // Runtime Loop
  //

  React.useEffect(() => {
    const serviceProgramStates = serviceProgramStatesRef.current

    // Loop function
    const task = async () => {
      // Programs that require immediate attention
      const urgentProgramIds = activeProgramIds.filter(programId => {
        const { state } = actionQueueMap[programId]
        // Don't execute programs which are assigned to another client.
        // This predicate must come first to avoid the device/client from
        // impacting program heing handled by other clients.
        if (state.clientId !== clientId) return false
        // Don't execute programs which are currently being executed by service
        if (state.executing && serviceProgramStates[programId]) {
          return false
        }
        // Detect program interruption if service is not evaluating the program
        // and program evaluation was executing an action op.
        if (state.executing && !serviceProgramStates[programId] && state.effective) {
          dispatch(
            updateActionProgramState({
              ...state,
              effect: { type: 'done', error: new Error('Program Interrupted') },
              executing: false
            })
          )
            // Use console.warn to not spam user
            .catch(err => console.warn(err))
          return false
        }
        // Don't execute programs which have not reached their scheduled time
        if (!state.effective && state.nextExecutionTime > Date.now()) return false
        // Otherwise, it's safe to execute
        return true
      })
      // Act on urgent programs
      const promises = urgentProgramIds.map(async programId => {
        const { program, state } = actionQueueMap[programId]

        // Set program state to executing
        await updateProgramState(state, true)

        // Use mock execution function if program is marked as mockMode
        const context = program.mockMode ? executionContextMock : executionContext

        const { nextState } = await executeActionProgram(context, program, state).catch((error): ExecutionResults => {
          const cleanError = asCleanError(error)
          console.warn(new Error('Action Program Exception: ' + cleanError.message))
          console.error(cleanError)
          return {
            nextState: {
              ...state,
              effect: { type: 'done', error: cleanError },
              effective: true,
              nextExecutionTime: -1
            }
          }
        })

        // Update program state
        await updateProgramState(nextState, false)
      })
      await Promise.all(promises)
    }

    // Setup loop
    const periodicTask = makePeriodicTask(task, EXECUTION_INTERVAL, {
      onError: error => {
        const errorMessage = String(error)
        console.warn(new Error('Unexpected error in ActionQueueService: ' + errorMessage))
        console.error(error)
      }
    })

    periodicTask.start()

    // Cleanup loop
    return () => periodicTask.stop()
  }, [account, clientId, dispatch, updateProgramState, actionQueueMap, activeProgramIds, executionContextMock, executionContext])

  // Return no component/view
  return null
}
