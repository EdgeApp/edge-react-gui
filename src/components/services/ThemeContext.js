// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
