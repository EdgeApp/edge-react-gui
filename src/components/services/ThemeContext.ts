import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
