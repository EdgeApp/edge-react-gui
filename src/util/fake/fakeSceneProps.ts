import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
