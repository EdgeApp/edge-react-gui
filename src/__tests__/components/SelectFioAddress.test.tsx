import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { getTheme } from '../../components/services/ThemeContext'
import { SelectFioAddressComponent } from '../../components/themed/SelectFioAddress'
import { fakeNavigation } from '../../util/fake/fakeSceneProps'

describe('SelectFioAddress', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const actual = renderer.render(
      <SelectFioAddressComponent
        navigation={fakeNavigation}
        selected="string"
        memo="TX"
        memoError="memo error"
        onSelect={() => undefined}
        onMemoChange={() => undefined}
        fioRequest={{
          fio_request_id: 123123,
          content: {
            payee_public_address: '123123',
            amount: '123123',
            token_code: 'LINK',
            chain_code: 'ETH',
            memo: 'myTX'
          },
          payee_fio_address: 'fio@fio',
          payer_fio_address: 'fio2@fio',
          payer_fio_public_key: '123123',
          status: 'requested',
          time_stamp: '11:11',
          fioWalletId: 'myWallet'
        }}
        currencyCode="BTC"
        isSendUsingFioAddress
        fioAddresses={[]}
        fioWallets={[]}
        selectedWalletId="123123"
        refreshAllFioAddresses={async () => {}}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
