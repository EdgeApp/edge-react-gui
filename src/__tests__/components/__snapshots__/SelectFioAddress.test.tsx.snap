// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectFioAddress should render with loading props 1`] = `
<View>
  <WithTheme(TileComponent)
    body="string"
    title="Send from FIO Crypto Handle"
    type="static"
  />
  <WithTheme(TileComponent)
    onPress={[Function]}
    title="FIO Memo Error"
    type="touchable"
  >
    <WithTheme(EdgeTextComponent)
      style={
        Object {
          "color": "#E85466",
        }
      }
    >
      memo error
    </WithTheme(EdgeTextComponent)>
  </WithTheme(TileComponent)>
</View>
`;
